add use cases

fforootd opened the pull request:
This is an initial commit containing the `useCases` and a first `apiDesign`

fforootd committed:
add job api

fforootd committed:
add useCases and basic microsService structure for the api

fforootd committed:
add usecases

fforootd committed:
add templates, dashboards, reports

fforootd committed:
add notification channel

adlerhurst commented on API/Jobs.proto:
```diff
@@ -0,0 +1,106 @@
+
+package yabs.jobs.api.v1;
+
+option go_package ="jobsAPI";
```
I prefer to call the package ´api ´

fforootd replied:
resolved with 21de16d4f3b281cf9516f1dbe0a22d3992cbcba5

adlerhurst commented on API/Jobs.proto:
```diff
@@ -0,0 +1,106 @@
+    // ---------
+    // Probes
+    // ---------
+    rpc Healthz(google.protobuf.Empty) returns (yabslabs.common.api.ReturnMessage) {
```
Maybe we can use ˋstatusˋ from 
https://cloud.google.com/apis/design/errors instead of ˋreturn message ˋ

fforootd replied:
i guess yes... what about the `validate`, is the `string message = 2;` sufficient for that use?

adlerhurst replied:
Validate shouldn‘t return status. Status is for error-messages. The returntype of validate is a list/map of dependencies

adlerhurst commented on API/Jobs.proto:
```diff
@@ -0,0 +1,106 @@
+    // ---------
+    rpc CreateNewJob() returns (Job) {
+        option (google.api.http) = {
+            get: "/Jobs"
```
I think ˋ/jobsˋ should be lower case

fforootd replied:
resolved with 21de16d4f3b281cf9516f1dbe0a22d3992cbcba5

adlerhurst commented on API/Jobs.proto:
```diff
@@ -0,0 +1,106 @@
+message Job {
+    uint64 jobid = 1; 
+    string name = 2; 
+    status status = 3;
```
Object type is upper case

adlerhurst replied:
Not upper case first letter should be big

fforootd replied:
resolved with 266abf6172b99145f9e482f26f1ccf97939a95a3

adlerhurst commented on API/Jobs.proto:
```diff
@@ -0,0 +1,106 @@
+    // ---------
+    // READ Jobs
+    // ---------
+    rpc GetAllJobs() returns (Jobs) {
```
If there is a use case to filter jobs i would rename this call to ˋgetJobs(Filters)ˋ

fforootd replied:
resolved with 266abf6172b99145f9e482f26f1ccf97939a95a3

adlerhurst commented on API/Jobs.proto:
```diff
@@ -0,0 +1,106 @@
+    // ---------
+    // UPDATE Jobs
+    // ---------
+    rpc UpdateJob(jobid) returns () {
```
Seperat ˋjobidˋ to ˋjob_idˋ

adlerhurst replied:
Sorry ˋjobidˋ has to be an object not seperated by underscore

hifabienne commented on UseCases/Groups.md:
```diff
@@ -0,0 +1,21 @@
+### Group Operations
+
+#### Add Group
+#### Add SubGroups
```
Remove Subgroup

fforootd replied:
added with commit 16a5d70805edf0e656ffffd0ccb489f2a86773dc

adlerhurst commented on UseCases/Jobs.md:
```diff
@@ -0,0 +1,16 @@
+
+#### Schedule Job
+#### Run Job
+
```
Abort job

fforootd replied:
added with commit 16a5d70805edf0e656ffffd0ccb489f2a86773dc

hifabienne commented on UseCases/Groups.md:
```diff
@@ -0,0 +1,21 @@
+#### Update Group
+#### Remove Group
+#### Deactivate Group
+#### Reactivate Group
```
Add Project
Remove Project

fforootd replied:
added with commit 16a5d70805edf0e656ffffd0ccb489f2a86773dc

fforootd committed:
add change from reviews

fforootd committed:
add review

fforootd committed:
add feedback from silvan

fforootd committed:
Merge branch 'master' into doc/api

fforootd committed:
update from review

fforootd committed:
Merge branch 'doc/api' of github.com:yabslabs/yabs into doc/api

stebenz commented on API/Jobs.proto:
```diff
@@ -0,0 +1,106 @@
+    // ---------
+    // CREATE Jobs
+    // ---------
+    rpc CreateNewJob() returns (Job) {
```
As a create action results in a new object I would just use "CreateJob"

fforootd replied:
resolved with 298a9fc6ba8b6e66888ea0a34cc9990603b5d5b0

stebenz commented on API/Jobs.proto:
```diff
@@ -0,0 +1,106 @@
+enum Status {
+    UNKNOWN = 0;
+    ADDED_NOT_VERIFIED = 1;
+    READY = 2;
```
Shouldn't status like FAILED and FINISHED also be included?

fforootd replied:
I added additional status infos

```
enum Status {
    UNKNOWN = 0;
    STUCK = 1;
    READY = 2;
    FAILED = 3;
    FINISHED = 4;
    PENDING = 5;
}
```

fforootd replied:
Gitlab has the following state for a `job`

[gitlab job api](https://docs.gitlab.com/ee/api/jobs.html)

i think we should add the same pattern

> Scope of jobs to show. Either one of or an array of the following: `created`, `pending`, `running`, `failed`, `success`, `canceled`, `skipped`, or `manual`. All jobs are returned if scope is not provided.

fforootd replied:
latest idea 

```
enum Status {
    UNKNOWN = 0;
    CREATED = 1;
    PENDING = 2;
    RUNNING = 3;
    FAILED = 4;
    SUCCESS = 5;
    CANCELED = 6;
    SKIPPED = 7;
    MANUAL = 8;
    STUCK = 9;
    RETRIED = 10;
    PAUSED = 11;
    SUSPENDED = 12;
}
```

adlerhurst replied:
Isn’t manual the same as created? A manual job has been created?
Success should be succeeded

fforootd replied:
i will remove the `manual` for now

fforootd committed:
changes from feedback

fforootd committed:
add feedback

fforootd committed:
add new status enums

fforootd committed:
new status codes

hifabienne commented on UseCases/Groups.md:
```diff
@@ -0,0 +1,24 @@
+
+#### List Group Members
+#### Add User to Group
+#### Remove User from Group
```
Some more lists
List Subgroups of Group
List Projects of Group

fforootd replied:
yes you are right

tribock committed:
remove -mod vendor

this flag has just to be used if a vendor folder is present in the root directory

tribock commented:
golangci-lint will continue to fail until there are some go packages in the project.
``` 
0.17s$ golangci-lint run
ERRO Running error: context loading failed: no go files to analyze 
The command "golangci-lint run" exited with 5.
```

fforootd committed:
latest feedbacks

fforootd commented:
i will merge this for now and create a new `MR` for the proto changes
